### PR TITLE
Do not write unnecessary VariableRecord.UPDATED if the variable remains unchanged

### LINKED ISSUE #7273: ProcessInstanceVariableTest#shouldUpdateMultipleVariables is flaky
**Summary**

- How often does the test fail? 3 times so far
- Does it block your work? not particularly
- Do we suspect that it is a real failure? no idea

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
Stacktrace
java.lang.AssertionError: 
Expecting ArrayList:
  [("x", "1"), ("z", "5")]
to contain:
  [("y", "4"), ("z", "5")]
but could not find the following element(s):
  [("y", "4")]
	at io.camunda.zeebe.engine.processing.variable.ProcessInstanceVariableTest.shouldUpdateMultipleVariables(ProcessInstanceVariableTest.java:321)
</pre>
</details>

**Logs**

<details><summary>Logs</summary>

```
Standard Output
12:29:41.120 [] INFO  io.camunda.zeebe.test - Test failed, following records were exported:
12:29:41.255 [] INFO  io.camunda.zeebe.test - Compact log representation:
--------
	['C'ommand/'E'event/'R'ejection] [valueType] [intent] - #[position]->#[source record position]  P[partitionId]K[key] - [summary of value]
	P9K999 - key; #999 - record position; "ID" element/process id; @"elementid"/[P9K999] - element with ID and key
	Keys are decomposed into partition id and per partition key (e.g. 2251799813685253 -> P1K005). If single partition, the partition is omitted.
	Long IDs are shortened (e.g. 'startEvent_5d56488e-0570-416c-ba2d-36d2a3acea78' -> 'star..acea78'
--------
E PROC_INST          COMPLETING     - #36->#35 K15 - START_EVENT "startEv..65e9fab" in <process "process"[K11]>
E PROC_INST          COMPLETED      - #37->#35 K15 - START_EVENT "startEv..65e9fab" in <process "process"[K11]>
E PROC_INST          SEQ_FLOW_TAKEN - #38->#35 K16 - SEQUENCE_FLOW "sequenc..7a1b671" in <process "process"[K11]>
C PROC_INST          ACTIVATE       - #39->#35 K17 - SERVICE_TASK "task" in <process "process"[K11]>
C PROC_INST_CREATION CREATE         - #40-> -1  -1 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          ACTIVATING     - #41->#39 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E JOB                CREATED        - #42->#39 K18 - K18 "test" @"task"[K17] 3 retries, in <process "process"[K11]> (no vars)
E PROC_INST          ACTIVATED      - #43->#39 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E VAR                CREATED        - #44->#40 K20 - x->1 in <process [K19]>
E VAR                CREATED        - #45->#40 K21 - z->3 in <process [K19]>
E VAR                CREATED        - #46->#40 K22 - y->2 in <process [K19]>
C PROC_INST          ACTIVATE       - #47->#40 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST_CREATION CREATED        - #48->#40 K23 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          ACTIVATING     - #49->#47 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #50->#47 K19 - PROCESS "process" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #51->#47  -1 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #52->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #53->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
C PROC_INST          COMPLETE       - #54->#51 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #55->#54 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #56->#54 K24 - START_EVENT "startEv..65e9fab" in <process "process"[K19]>
E PROC_INST          SEQ_FLOW_TAKEN - #57->#54 K25 - SEQUENCE_FLOW "sequenc..7a1b671" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #58->#54 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #59->#58 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E JOB                CREATED        - #60->#58 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K19]> (no vars)
E PROC_INST          ACTIVATED      - #61->#58 K26 - SERVICE_TASK "task" in <process "process"[K19]>
C JOB                COMPLETE       - #62-> -1 K27 - K27 "test" @""[ -1] -1 retries, in <process ?[?]> with variables: {x=1, y=4, z=5}
E JOB                COMPLETED      - #63->#62 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K19]> with variables: {x=1, y=4, z=5}
C PROC_INST          COMPLETE       - #64->#62 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #65->#64 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E VAR                UPDATED        - #66->#64 K20 - x->1 in <process [K19]>
E VAR                UPDATED        - #67->#64 K21 - z->5 in <process [K19]>
E VAR                UPDATED        - #68->#64 K22 - y->4 in <process [K19]>
E PROC_INST          COMPLETED      - #69->#64 K26 - SERVICE_TASK "task" in <process "process"[K19]>
E PROC_INST          SEQ_FLOW_TAKEN - #70->#64 K28 - SEQUENCE_FLOW "sequenc..f9809c2" in <process "process"[K19]>
C PROC_INST          ACTIVATE       - #71->#64 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          ACTIVATING     - #72->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          ACTIVATED      - #73->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #74->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #75->#71 K29 - END_EVENT "endEven..c88755c" in <process "process"[K19]>
C PROC_INST          COMPLETE       - #76->#71 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          COMPLETING     - #77->#76 K19 - PROCESS "process" in <process "process"[K19]>
E PROC_INST          COMPLETED      - #78->#76 K19 - PROCESS "process" in <process "process"[K19]>
-------------- Deployed Processes ----------------------
--------------- Decomposed keys (for debugging) -----------------
 -1 <-> -1
K11 <-> 2251799813685259
K15 <-> 2251799813685263
K16 <-> 2251799813685264
K17 <-> 2251799813685265
K18 <-> 2251799813685266
K19 <-> 2251799813685267
K20 <-> 2251799813685268
K21 <-> 2251799813685269
K22 <-> 2251799813685270
K23 <-> 2251799813685271
K24 <-> 2251799813685272
K25 <-> 2251799813685273
K26 <-> 2251799813685274
K27 <-> 2251799813685275
K28 <-> 2251799813685276
K29 <-> 2251799813685277
```

</details>

### LINKED ISSUE #7295: Unnecessary `VariableRecord.UPDATED` if the variable exists and its value has not changed
**Describe the bug**

We currently write `VariableRecord.UPDATED` if the variable exists, even if the value has not changed. This is a bug introduced during the engine refactoring - prior to this, we omitted such records. Unfortunately our tests did not catch the change, and only updating a dependency (which then shuffled the order in which the variables are changed) caused a test to start failing.

**To Reproduce**

Update `agrona` to 1.11, then run `ProcessInstanceVariableTest#shouldUpdateMultipleVariables`.

**Expected behavior**

We only publish a `VariableRecord.UPDATED` if the variable has changed.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
C PROC_INST_CREATION CREATE         - #36-> -1  -1 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          COMPLETING     - #37->#35 K15 - START_EVENT "startEv..077ef91" in <process "process"[K11]>
E PROC_INST          COMPLETED      - #38->#35 K15 - START_EVENT "startEv..077ef91" in <process "process"[K11]>
E PROC_INST          SEQ_FLOW_TAKEN - #39->#35 K16 - SEQUENCE_FLOW "sequenc..35e84ff" in <process "process"[K11]>
C PROC_INST          ACTIVATE       - #40->#35 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E VAR                CREATED        - #41->#36 K19 - x->1 in <process [K18]>
E VAR                CREATED        - #42->#36 K20 - z->3 in <process [K18]>
E VAR                CREATED        - #43->#36 K21 - y->2 in <process [K18]>
C PROC_INST          ACTIVATE       - #44->#36 K18 - PROCESS "process" in <process "process"[K18]>
E PROC_INST_CREATION CREATED        - #45->#36 K22 - new <process "process"> with variables: {x=1, y=2, z=3}
E PROC_INST          ACTIVATING     - #46->#40 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E JOB                CREATED        - #47->#40 K23 - K23 "test" @"task"[K17] 3 retries, in <process "process"[K11]> (no vars)
E PROC_INST          ACTIVATED      - #48->#40 K17 - SERVICE_TASK "task" in <process "process"[K11]>
E PROC_INST          ACTIVATING     - #49->#44 K18 - PROCESS "process" in <process "process"[K18]>
E PROC_INST          ACTIVATED      - #50->#44 K18 - PROCESS "process" in <process "process"[K18]>
C PROC_INST          ACTIVATE       - #51->#44  -1 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
E PROC_INST          ACTIVATING     - #52->#51 K24 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
E PROC_INST          ACTIVATED      - #53->#51 K24 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
C PROC_INST          COMPLETE       - #54->#51 K24 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
E PROC_INST          COMPLETING     - #55->#54 K24 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
E PROC_INST          COMPLETED      - #56->#54 K24 - START_EVENT "startEv..077ef91" in <process "process"[K18]>
E PROC_INST          SEQ_FLOW_TAKEN - #57->#54 K25 - SEQUENCE_FLOW "sequenc..35e84ff" in <process "process"[K18]>
C PROC_INST          ACTIVATE       - #58->#54 K26 - SERVICE_TASK "task" in <process "process"[K18]>
E PROC_INST          ACTIVATING     - #59->#58 K26 - SERVICE_TASK "task" in <process "process"[K18]>
E JOB                CREATED        - #60->#58 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K18]> (no vars)
E PROC_INST          ACTIVATED      - #61->#58 K26 - SERVICE_TASK "task" in <process "process"[K18]>
C JOB                COMPLETE       - #62-> -1 K27 - K27 "test" @""[ -1] -1 retries, in <process ?[?]> with variables: {x=1, y=4, z=5}
E JOB                COMPLETED      - #63->#62 K27 - K27 "test" @"task"[K26] 3 retries, in <process "process"[K18]> with variables: {x=1, y=4, z=5}
C PROC_INST          COMPLETE       - #64->#62 K26 - SERVICE_TASK "task" in <process "process"[K18]>
E PROC_INST          COMPLETING     - #65->#64 K26 - SERVICE_TASK "task" in <process "process"[K18]>
E VAR                UPDATED        - #66->#64 K19 - x->1 in <process [K18]>
E VAR                UPDATED        - #67->#64 K20 - z->5 in <process [K18]>
E VAR                UPDATED        - #68->#64 K21 - y->4 in <process [K18]>
E PROC_INST          COMPLETED      - #69->#64 K26 - SERVICE_TASK "task" in <process "process"[K18]>
E PROC_INST          SEQ_FLOW_TAKEN - #70->#64 K28 - SEQUENCE_FLOW "sequenc..dff4c5f" in <process "process"[K18]>
C PROC_INST          ACTIVATE       - #71->#64 K29 - END_EVENT "endEven..6cfdc62" in <process "process"[K18]>
E PROC_INST          ACTIVATING     - #72->#71 K29 - END_EVENT "endEven..6cfdc62" in <process "process"[K18]>
E PROC_INST          ACTIVATED      - #73->#71 K29 - END_EVENT "endEven..6cfdc62" in <process "process"[K18]>
E PROC_INST          COMPLETING     - #74->#71 K29 - END_EVENT "endEven..6cfdc62" in <process "process"[K18]>
E PROC_INST          COMPLETED      - #75->#71 K29 - END_EVENT "endEven..6cfdc62" in <process "process"[K18]>
C PROC_INST          COMPLETE       - #76->#71 K18 - PROCESS "process" in <process "process"[K18]>
E PROC_INST          COMPLETING     - #77->#76 K18 - PROCESS "process" in <process "process"[K18]>
E PROC_INST          COMPLETED      - #78->#76 K18 - PROCESS "process" in <process "process"[K18]>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR fixes a bug introduced during the engine refactoring, where we now write `VariableRecord.UPDATED` even if the variable remains unchanged. The fix removes this unnecessary follow up event. This is technically a breaking change, insofar as applications relying on this record being produced would not find it anymore. However, there is no loss of information - the variable CREATED still exist, and the command which would have produced the unnecessary `VariableRecord.UPDATED` is still exported and has all the info required, so I think it's a change we can accept.

## Related issues

closes #7295 
closes #7273

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by korthout — APPROVED** [review]
Generally a great addition, thanks for these changes!

Please resolve the issues I mentioned before merging

File: engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
Comment (npepinpe) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java:358]: True, I'll rename it :+1:

File: engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java
Comment (npepinpe) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java:317]: Oh yeah, I missed that :sweat_smile:

File: engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java
Comment (npepinpe) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java:225]: AFAIK `satisfiesExactlyInAnyOrder` compares the length of the iterable with the number of consumers, so it wouldn't pass if the list was empty.

Comment (npepinpe) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java:225]: Well there's an easy to know, I'll just test it out :sweat_smile:

Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/variable/VariableBehaviorTest.java:225]: Nevermind, I should read docs before making such statements 😆 Thanks for looking into it

**Review by korthout — COMMENTED** [review]
Seems that the `shouldUseTheSameLoopVariablesWhenIncidentResolved` test case still takes above 5 seconds now. It's not directly clear to me why it takes so long though.

**Comment by npepinpe** [comment]
Because the incident is never resolved correctly and the process never completes :smile:

**Comment by npepinpe** [comment]
Was it intended that the incident would not be resolved?

File: engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java:307]: This does not look deterministic, although I might be wrong here.

Please also verify that the variable was updated before resolving the incidents.

**Review by korthout — APPROVED** [review]
>Was it intended that the incident would not be resolved?

This was never part of the test, it didn't matter. I like that we resolve them now, albeit a little outside of the scope of this test.
Please consider my last comment before merging

File: engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java
Comment (npepinpe) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java:307]: The variable is updated as part of the command processing, as in, the state is modified. `update()` returns the response from the client, so I would assume it's deterministic, as in the variable is 100% updated when we receive the response. See the success expectation:

```java
private static final LongFunction<Record<VariableDocumentRecordValue>>
      SUCCESSFUL_EXPECTATION_SUPPLIER =
          (sourceRecordPosition) ->
              RecordingExporter.variableDocumentRecords(VariableDocumentIntent.UPDATED)
                  .withSourceRecordPosition(sourceRecordPosition)
                  .getFirst();
```

And the update method:

```java
public Record<VariableDocumentRecordValue> update() {
    final long position =
        environmentRule.writeCommand(VariableDocumentIntent.UPDATE, variableDocumentRecord);
    return expectation.apply(position);
  }
```

I simply ignore the return value here, but I think we can safely say the variable was updated. Here are the records produced by updating a document:

```
C VAR_DOCUMENT       UPDATE         - #040->  -1   -1 - VariableDocumentRecord {"variables":{"y":1},"updateSemantics":"PROPAGATE","scopeKey":2251799813685253}
E VAR                CREATED        - #041->#040 K023 - y->1 in <process [K005]>
E VAR_DOCUMENT       UPDATED        - #042->#040 K024 - VariableDocumentRecord {"variables":{"y":1},"updateSemantics":"PROPAGATE","scopeKey":2251799813685253}
```

Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java:307]: I was unaware that we wait for the response. This makes sense. Thanks for the clear description :)

**Comment by npepinpe** [comment]
bors merge

**Comment by npepinpe** [comment]
bors retry

**Pull request merged**
